pkg/controller/cluster/cluster_controller.go

```diff
@@ -119,18 +142,60 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
 	if !cluster.ObjectMeta.DeletionTimestamp.IsZero() {
 		// no-op if finalizer has been removed.
 		if !util.Contains(cluster.ObjectMeta.Finalizers, clusterv1.ClusterFinalizer) {
-			klog.Infof("reconciling cluster object %v causes a no-op as there is no finalizer.", name)
+			klog.Infof("Reconciling cluster object %v causes a no-op as there is no finalizer.", name)
 			return reconcile.Result{}, nil
 		}
 
-		klog.Infof("reconciling cluster object %v triggers delete.", name)
+		children, err := r.listChildren(context.Background(), cluster)
+		if err != nil {
+			klog.Infof("Failed to list dependent objects of cluster %s/%s: %v", cluster.ObjectMeta.Namespace, cluster.ObjectMeta.Name, err)
+			return reconcile.Result{}, errors.WithStack(err)
+		}
+
+		if len(children) > 0 {
+			klog.Infof("Deleting cluster %s: %d children still exist, will requeue", name, len(children))
+
+			var errList []error
+
+			for _, child := range children {
+				accessor, err := meta.Accessor(child)
+				if err != nil {
+					return reconcile.Result{}, errors.Wrapf(err, "couldn't create accessor for %T", child)
+				}
+
+				if accessor.GetDeletionTimestamp() != nil {
+					continue
+				}
+
+				gvk := child.GetObjectKind().GroupVersionKind().String()
+
+				klog.V(4).Infof("Deleting cluster %s: Deleting %s %s", name, gvk, accessor.GetName())
+				if err := r.Delete(context.Background(), child, client.PropagationPolicy(metav1.DeletePropagationForeground)); err != nil {
+					err = errors.Wrapf(err, "deleting cluster %s: failed to delete %s %s", name, gvk, accessor.GetName())
+					klog.Infof("deletion error: %v", err)
+					errList = append(errList, err)
+				}
+			}
+
+			if len(errList) == 1 {
```

Comment:
The Aggregate code handles array lengths of 0, 1, and 2+ correctly - you can safely replace it with a check for > 0.